src/client/datascience/kernel-launcher/localPythonAndRelatedNonPythonKernelSpecFinder.ts

```diff
@@ -149,7 +128,7 @@ export class LocalPythonAndRelatedNonPythonKernelSpecFinder extends LocalKernelS
         // Go through the global kernelspecs that use python to launch the kernel but don't use ipykernel.
         await Promise.all(
             globalKernelSpecs
```

Comment:
So this check for ipykernel_launcher on global python spec was what was hiding my kernelspecs. This kernel finding code has gotten a bit twisty, so I want to make sure that I'm not regression something with this change. But I can't see a reason why this check was here before. I can't see a reason to hide these specs from the user.

Reply:
The reason was global ipykernel specs- one's that match an interpreter should be skipped because the 'interpreter' kernelspec would find them.

Reply:
We don't want doubles.

Reply:
By "interpreter" kernelspec do you mean the kernelspecs that we create for each interpreter? If so that doesn't work, it's very easy to have multiple global kernelspecs with different names and environments pointing at the same interpreter.

Reply:
They can have different names but what if they're the same. From the user's point of view they can't tell which one is which (as the python path is the same) and they're functionally equivalent (well unless the env is different for each).

Reply:
The different envs is exactly it. My repro has global kernelspecs with different envs pointing at the same interpreter.

Reply:
I think a better way to handle this is to display global python kernel specs ONLY if they are not craeted by us and if they have custom env variables.

That way we avoid duplicates & go back to the previous approach (with the exception of dipslaying global kernelspecs with custom env variables)

Reply:
@DonJayamanne I believe that would work, but checking to make sure now. That would be pretty similar to what the non-global kernels is already doing here:
https://github.com/microsoft/vscode-jupyter/blob/b7241b5b97c9e0ff4ab7d4aa26c0d1d0b9a643e0/src/client/datascience/kernel-launcher/localPythonAndRelatedNonPythonKernelSpecFinder.ts#L204

Reply:
Aaah yes, that's correct that should be sufficient.

Reply:
Actually, I think we should modify the code to return all python kernels except those with custom env, just like the code you highlighted.
Else i don't think it will work.
Will wait for your confirmation.

Reply:
I was just playing with the condition now for global python spec inclusion as:
!isKernelRegisteredByUs && (usingNonIpyKernel || customEnv)  // plus it needs to match an interpreter
For the globalKernelSpecs

This works for my scenario, but it does reverse what is happening in the non global case that I pointed out. Basically the question in my head is if it matters if the kernel spec hides the interpreter as kernel or visa-versa (plus I was starting to add a test for this scenario).

Reply:
> Basically the question in my head is if it matters if the kernel spec hides the interpreter as kernel 

I don' tthink it should, 
Also I'd expect the names to be different (i.e. when users create the names, I do'nt expect the names to be exactly the same as the display names we use for interpreters, which contians python name, version, etc)

Reply:
I feel like there is a lot of "de-duplication" stuff going on for the kernel spec finding, but it's a bit spread out all over the place. I'm almost tempted to consider making all the finding stuff just return flat lists, then have a final single deduplication step to sort it all out. That would be a pretty big change though, and I don't like the idea of perturbing this now that it's in pretty good shape now.

Reply:
> I'm almost tempted to consider making all the finding stuff just return flat lists, then have a final single deduplication step to sort it all out

Sounds great, right now it does feel like the code could use some cleanup

> nd I don't like the idea of perturbing this now that it's in pretty good shape now.

Meaning, this is complicated code, i think we should try to clean it up in debt or when ever, if we're concerned about making changes, that just re-iterates the fact that its not in the best shape & not easy to maintain.

But totally fine not refactoring it now.

Reply:
Our kernel finding has improved a ton recently, just a complicated problem, as you mentioned. I just find myself thinking "I'm preeeeetty sure this won't break anything" whenever I'm making changes in this area. Low confidence about regressions or side effects.

Reply:
+1 on the filter in one spot. Sounds like a good refactor to me too. Not sure I'd do it now but maybe next iteration.